### ISSUE TITLE
Implementing an arbitrary base distribution within the normalising flow

### ISSUE THREAD

**Issue by aiwe2320** [E471r/FEcrys#4]
Hello,

We have an interest in modifying the base distribution such that the normalizing flow maps between the target crystal distribution and Gaussian noise, as opposed to a latent uniform distribution. 

As a note, we are considering the use of FEcrys from a heavily simplified perspective:
1) We are only considering monatomic Lennard Jones crystals, which simplifies the representation and training procedure greatly
2) We are not using the FEcrys infrastructure to evaluate the free energies of our subject crystals. Practically, this means we are running the training procedure with `evaluate_main = False` and are not concerned with generating free energy estimates during training at the moment.

With these simplifications in mind, do you have any thoughts as to where in the FEcrys code we should start in order to change the base distribution?  What information within the normalizing flow architecture would we need to change in order to make this possible?

My current understanding of the current normalizing flow code is that this would primarily be a change to the training process. In particular, the base probability density function `ln_base_` would need to be changed to reflect the new latent distribution. I am focusing on the function `NN.pgm.model_helper_PGMcrys_v1.step_ML_graph` as the primary target for altering the base distribution, but I am struggling to fully understand the function nesting within `self._forward_coupling_(x)`. Would altering the definition of the base probability density be sufficient to change the model's mapping, or are there any other parts of the training process that we would need to change? 

I apologize if this question is naive regarding your repository, normalizing flows, or both. Please let us know if you have any thoughts on the relative difficulty of learning to an arbitrary base distribution using the current infrastructure. Thank you for your time.

**Comment by E471r** [comment]
Hello,

Ill jump straight to this:
The PGMcrys_v1 class, or any similar class (the self = model), during the __init__ step has these two lines, which can be changed to any another base distribution p_{0}:

self.ln_base_ = self.ic_maps[0].ln_base_
self.sample_base_ = self.ic_maps[0].sample_base_

^This is the answer (you can try this already), but here are further details:

# please find self.ic_maps[0].ln_base_ defined inside SingleComponent_map
# please find self.ic_maps[0].sample_base_ defined inside SingleComponent_map

The two functions from above (self.ln_base_, self.sample_base_) are then used inside (self.ln_model, self.ln_model_for_step_ML_) and in self.sample_model, respectively. These are inherited from model_helper. Im sorry they are scattered over 3 different files. You can move many of the functions around into a single (model) class to see all the methods in one file.

The quickest modification to swap from a uniform p_{0} (default) to something else is by defining your own version of base-distribution-related components and putting them inside the __init__ step (i.e., replacing the two lines to something like):

self.ln_base_ = lambda z : ...          # output: (m,1) array of log probabilities (\ln{p_{0}})
self.sample_base_= lambda m : ... # output: same shape(s) as z

In the above, z is an array, or a list if arrays, made by z = _forward_coupling_(x), where x = _forward_represenation_(r), where r Cartesian coordinates.
In the above, m is integer describing the batch size to sample (number of supercells to sample). 

Please use truncated distributions between [-1,1] or [-1,1) for the custom base distribution. The package tensorflow-probability (tfp) has useful truncated distributions (like tfp.distributions.TruncatedNormal, or tfp.distributions.VonMises, respectively). Please let me know I can add a p_{0} to the repository that may helpful you. I did not add any so far because I prefer uniform, because the splines are expressive enough, and it is slightly cheaper. 

Let me know how it goes.
